File: frontend/static/js/pages/attendance.js
// Attendance Page
let _attPage = 1;

async function renderAttendance() {
  const today = new Date().toISOString().split('T')[0];
  document.getElementById('page-content').innerHTML = `
    <div class="page-topbar">
      <div class="page-title-block">
        <h1 class="page-title">Attendance</h1>
        <p class="page-subtitle">Track check-ins, check-outs, and work hours</p>
      </div>
      <div class="page-actions">
        <button class="btn btn-secondary btn-sm" id="exportAttBtn" onclick="exportAtt()">Export CSV</button>
        <button class="btn btn-primary" onclick="openManualCheckin()">+ Manual Check-in</button>
      </div>
    </div>
    <div class="page-body">
      <div class="stat-grid" id="attStats">
        ${[1,2,3,4].map(()=>`<div class="stat-card"><div class="skeleton" style="height:70px"></div></div>`).join('')}
      </div>

      <div class="toolbar">
        <div class="form-group" style="flex-direction:row;align-items:center;gap:8px;margin:0">
          <label class="form-label" style="white-space:nowrap">DATE</label>
          <input type="date" class="form-control" id="attDate" value="${today}" style="width:160px" onchange="loadAttTable()"/>
        </div>
        <div class="search-box">
          <input type="text" id="attSearch" placeholder="Search employee..." oninput="loadAttTable()"/>
        </div>
        <select class="form-control" id="attDeptFilter" style="width:170px" onchange="loadAttTable()">
          <option value="">All Departments</option>
        </select>
        <select class="form-control" id="attStatusFilter" style="width:140px" onchange="loadAttTable()">
          <option value="">All</option>
          <option value="present">Present</option>
          <option value="absent">Absent</option>
        </select>
        <div class="toolbar-spacer"></div>
        <span class="tag tag-muted" id="attCountTag">Loading...</span>
      </div>

      <div class="card">
        <div class="table-scroll">
          <table class="data-table" role="grid" aria-label="Attendance records">
            <thead>
              <tr>
                <th>Employee</th>
                <th>Department</th>
                <th>Check In</th>
                <th>Check Out</th>
                <th>Duration</th>
                <th>Method</th>
                <th>Confidence</th>
                <th>Status</th>
                <th>Actions</th>
              </tr>
            </thead>
            <tbody id="attTableBody">${loadingRows(9,8)}</tbody>
          </table>
        </div>
        <div id="attPagination"></div>
      </div>
    </div>`;

  await loadAttDepts();
  loadAttStats();
  loadAttTable();
}

async function loadAttDepts() {
  try {
    const depts = await API.departments.list();
    const sel = document.getElementById('attDeptFilter');
    if(sel) sel.innerHTML = '<option value="">All Departments</option>' +
      depts.map(d=>`<option value="${d.id}">${d.name}</option>`).join('');
  } catch(e) {}
}

async function loadAttStats() {
  try {
    const stats = await API.employees.stats();
    document.getElementById('attStats').innerHTML = `
      <div class="stat-card">
        <div class="stat-card-accent" style="background:var(--periwinkle)"></div>
        <div class="stat-label">Today's Date</div>
        <div class="stat-value" style="font-size:18px;margin-top:8px;font-family:var(--font-mono);color:var(--periwinkle)">${new Date().toLocaleDateString('en-UG',{day:'2-digit',month:'short',year:'numeric'})}</div>
      </div>
      <div class="stat-card">
        <div class="stat-card-accent" style="background:var(--mint)"></div>
        <div class="stat-label">Present</div>
        <div class="stat-value" style="color:var(--mint)">${stats.present_today}</div>
        <div class="stat-delta up">of ${stats.total_employees} employees</div>
      </div>
      <div class="stat-card">
        <div class="stat-card-accent" style="background:var(--coral)"></div>
        <div class="stat-label">Absent</div>
        <div class="stat-value" style="color:var(--coral)">${stats.absent_today}</div>
        <div class="stat-delta dn">${100-stats.attendance_rate}% absence</div>
      </div>
      <div class="stat-card">
        <div class="stat-card-accent" style="background:var(--citron)"></div>
        <div class="stat-label">Attendance Rate</div>
        <div class="stat-value" style="color:var(--citron)">${stats.attendance_rate}%</div>
        <div style="margin-top:6px"><div class="conf-bar"><div class="conf-fill" style="width:${stats.attendance_rate}%;background:var(--citron)"></div></div></div>
      </div>`;
  } catch(e) {}
}

async function loadAttTable(page=1) {
  _attPage = page;
  const date   = document.getElementById('attDate')?.value || new Date().toISOString().split('T')[0];
  const q      = document.getElementById('attSearch')?.value || '';
  const dept   = document.getElementById('attDeptFilter')?.value || '';
  const status = document.getElementById('attStatusFilter')?.value || '';
  try {
    const data = await API.attendance.list({ date, q, department_id:dept, status, page, per_page:20 });
    document.getElementById('attCountTag').textContent = `${data.total} records`;

    const tbody = document.getElementById('attTableBody');
    if(!data.records.length) {
      tbody.innerHTML=`<tr><td colspan="9"><div class="empty-state"><div class="empty-icon">o</div>No records found</div></td></tr>`;
      return;
    }
    tbody.innerHTML = data.records.map(r => {
      const statusBadge = r.is_present
        ? (r.check_out ? '<span class="tag tag-mint">Complete</span>' : '<span class="tag tag-citron">Active</span>')
        : '<span class="tag tag-coral">Absent</span>';
      const methodTag = r.method==='face'?'<span class="tag tag-peri">Face</span>':'<span class="tag tag-muted">Manual</span>';
      const conf = r.confidence ? `<span style="font-family:var(--font-mono);font-size:12px;color:${r.confidence>85?'var(--mint)':'var(--amber)'}">${r.confidence}%</span>` : '—';
      const dur = fmtDuration(r.check_in, r.check_out);
      return `<tr>
        <td>
          <div style="display:flex;align-items:center;gap:10px">
            ${avatar(r.name)}
            <div>
              <div style="font-weight:600">${r.name}</div>
              <div style="font-size:11px;color:var(--text-2);font-family:var(--font-mono)">${r.employee_no||'—'}</div>
            </div>
          </div>
        </td>
        <td><span class="tag tag-muted" style="border-left:3px solid ${r.dept_color||'var(--border2)'}">${r.department||'—'}</span></td>
        <td><span style="font-family:var(--font-mono);font-size:12px;color:var(--mint)">${fmtTime(r.check_in)}</span></td>
        <td><span style="font-family:var(--font-mono);font-size:12px;color:var(--amber)">${fmtTime(r.check_out)}</span></td>
        <td><span style="font-family:var(--font-mono);font-size:12px;color:var(--text-2)">${dur}</span></td>
        <td>${methodTag}</td>
        <td>${conf}</td>
        <td>${statusBadge}</td>
        <td>
          <div style="display:flex;gap:6px">
            ${r.record_id
              ? `<button class="btn btn-ghost btn-sm" onclick="quickCheckout(${r.record_id})">Check Out</button>
                 <button class="btn btn-danger btn-sm" onclick="deleteAtt(${r.record_id})">x</button>`
              : `<button class="btn btn-secondary btn-sm" onclick="quickCheckinEmp(${r.employee_id})">Check In</button>`}
          </div>
        </td>
      </tr>`;
    }).join('');
    document.getElementById('attPagination').innerHTML = paginate(data.total, page, 20, 'loadAttTable');
  } catch(e) { toast('Failed: '+e.message,'error'); }
}

async function quickCheckinEmp(empId) {
  try {
    const r = await API.attendance.checkin({ employee_id:empId, method:'manual', confidence:100 });
    toast('Checked in','success'); loadAttTable(_attPage); loadAttStats();
  } catch(e) { toast('Error: '+e.message,'error'); }
}

async function quickCheckout(recordId) {
  try {
    await API.attendance.update(recordId, { check_out: new Date().toISOString() });
    toast('Checked out','success'); loadAttTable(_attPage);
  } catch(e) { toast('Error: '+e.message,'error'); }
}

async function deleteAtt(recordId) {
  const ok = await confirmDialog('Delete this attendance record? This action cannot be undone.', true);
  if(!ok) return;
  try { await API.attendance.delete(recordId); toast('Record deleted','info'); loadAttTable(_attPage); loadAttStats(); }
  catch(e) { toast('Error: '+e.message,'error'); }
}

function exportAtt() {
  const date = document.getElementById('attDate')?.value || new Date().toISOString().split('T')[0];
  window.open(API.attendance.exportUrl(date));
}

async function openManualCheckin() {
  try {
    const empData = await API.employees.list({ per_page:200, status:'active' });
    openModal(`
      <div class="modal-header"><span class="modal-title">Manual Check-in</span><button class="modal-close" onclick="closeModal()">x</button></div>
      <div class="modal-body">
        <div class="form-group">
          <label class="form-label">Employee *</label>
          <select class="form-control" id="manEmp"><option value="">Choose employee...</option>
            ${empData.employees.map(e=>`<option value="${e.id}">${e.first_name} ${e.last_name} — ${e.department_name||'N/A'}</option>`).join('')}
          </select>
        </div>
        <div class="form-group">
          <label class="form-label">Check-in Time</label>
          <input class="form-control" id="manTime" type="datetime-local" value="${new Date().toISOString().slice(0,16)}"/>
        </div>
        <div class="form-group">
          <label class="form-label">Notes</label>
          <input class="form-control" id="manNotes" placeholder="Optional reason..."/>
        </div>
      </div>
      <div class="modal-footer">
        <button class="btn btn-secondary" onclick="closeModal()">Cancel</button>
        <button class="btn btn-primary" onclick="submitManualCheckin()">Confirm Check-in</button>
      </div>`);
  } catch(e) {}
}

async function submitManualCheckin() {
  const empId = document.getElementById('manEmp').value;
  if(!empId) { toast('Select employee','error'); return; }
  try {
    await API.attendance.checkin({ employee_id:parseInt(empId), method:'manual', confidence:100 });
    toast('Manual check-in recorded','success');
    closeModal(); loadAttTable(_attPage); loadAttStats();
  } catch(e) { toast('Error: '+e.message,'error'); }
}

File: backend/routes/attendance.py
from auth_utils import login_required, role_required
from flask import Blueprint, jsonify, request
from database import get_db
from datetime import datetime, date

attendance_bp = Blueprint('attendance', __name__)

@attendance_bp.route('/', methods=['GET'])
@login_required
def list_attendance():
    db = get_db()
    att_date = request.args.get('date', date.today().strftime('%Y-%m-%d'))
    dept = request.args.get('department_id', '')
    status = request.args.get('status', '')
    q = request.args.get('q', '')
    page = int(request.args.get('page', 1))
    per_page = int(request.args.get('per_page', 50))

    sql = """
        SELECT e.id AS employee_id, e.first_name||' '||e.last_name AS name,
               e.employee_no, e.position, d.name AS department, d.color AS dept_color,
               a.id AS record_id, a.check_in, a.check_out, a.status,
               a.confidence, a.method, a.overtime_hrs, a.notes,
               CASE WHEN a.check_in IS NOT NULL THEN 1 ELSE 0 END AS is_present
        FROM employees e
        LEFT JOIN departments d ON d.id = e.department_id
        LEFT JOIN attendance a ON a.employee_id = e.id AND a.date = ?
        WHERE e.status = 'active'
    """
    params = [att_date]
    if dept:
        sql += " AND e.department_id = ?"
        params.append(dept)
    if status == 'present':
        sql += " AND a.id IS NOT NULL"
    elif status == 'absent':
        sql += " AND a.id IS NULL"
    if q:
        sql += " AND (e.first_name||' '||e.last_name LIKE ? OR e.employee_no LIKE ?)"
        params += [f'%{q}%', f'%{q}%']

    total = db.execute(f"SELECT COUNT(*) FROM ({sql})", params).fetchone()[0]
    sql += f" ORDER BY is_present DESC, e.first_name LIMIT {per_page} OFFSET {(page-1)*per_page}"
    rows = db.execute(sql, params).fetchall()
    db.close()
    return jsonify({
        'records': [dict(r) for r in rows],
        'total': total, 'date': att_date
    })

@attendance_bp.route('/checkin', methods=['POST'])
def checkin():
    db = get_db()
    data = request.json
    emp_id = data.get('employee_id')
    method = data.get('method', 'manual')
    confidence = data.get('confidence', 100)
    today = date.today().strftime('%Y-%m-%d')
    now = datetime.now().isoformat(timespec='seconds')

    existing = db.execute("SELECT * FROM attendance WHERE employee_id=? AND date=?",
                          (emp_id, today)).fetchone()
    if existing:
        if existing['check_out']:
            return jsonify({'message': 'Already completed for today', 'type': 'done'}), 200
        # Check out
        db.execute("UPDATE attendance SET check_out=?, updated_at=datetime('now') WHERE id=?",
                   (now, existing['id']))
        db.commit(); db.close()
        return jsonify({'message': 'Checked out', 'type': 'checkout', 'time': now})

    db.execute("""
        INSERT INTO attendance (employee_id,date,check_in,method,confidence,status)
        VALUES (?,?,?,?,?,'present')
    """, (emp_id, today, now, method, confidence))
    db.commit(); db.close()
    return jsonify({'message': 'Checked in', 'type': 'checkin', 'time': now})

@attendance_bp.route('/<int:record_id>', methods=['PUT'])
def update_record(record_id):
    db = get_db()
    data = request.json
    fields = ['check_in','check_out','status','notes','overtime_hrs']
    sets = ', '.join(f"{f}=?" for f in fields if f in data)
    vals = [data[f] for f in fields if f in data] + [record_id]
    db.execute(f"UPDATE attendance SET {sets} WHERE id=?", vals)
    db.commit(); db.close()
    return jsonify({'message': 'Updated'})

@attendance_bp.route('/<int:record_id>', methods=['DELETE'])
def delete_record(record_id):
    db = get_db()
    db.execute("DELETE FROM attendance WHERE id=?", (record_id,))
    db.commit(); db.close()
    return jsonify({'message': 'Deleted'})

@attendance_bp.route('/history/<int:emp_id>', methods=['GET'])
def employee_history(emp_id):
    db = get_db()
    rows = db.execute("""
        SELECT date, check_in, check_out, status, confidence, method, overtime_hrs
        FROM attendance WHERE employee_id=? ORDER BY date DESC LIMIT 30
    """, (emp_id,)).fetchall()
    db.close()
    return jsonify([dict(r) for r in rows])

@attendance_bp.route('/export', methods=['GET'])
def export_csv():
    from flask import Response
    db = get_db()
    att_date = request.args.get('date', date.today().strftime('%Y-%m-%d'))
    rows = db.execute("""
        SELECT e.employee_no, e.first_name||' '||e.last_name AS name,
               d.name AS department, e.position,
               a.check_in, a.check_out, a.status, a.confidence, a.overtime_hrs
        FROM employees e
        LEFT JOIN departments d ON d.id = e.department_id
        LEFT JOIN attendance a ON a.employee_id = e.id AND a.date = ?
        WHERE e.status = 'active'
        ORDER BY d.name, e.first_name
    """, (att_date,)).fetchall()
    db.close()
    lines = ['Employee No,Name,Department,Position,Check In,Check Out,Status,Confidence,Overtime Hrs']
    for r in rows:
        def t(v): return str(v or '')
        ci = r['check_in'].split('T')[1][:5] if r['check_in'] else ''
        co = r['check_out'].split('T')[1][:5] if r['check_out'] else ''
        lines.append(f"{t(r['employee_no'])},{t(r['name'])},{t(r['department'])},{t(r['position'])},{ci},{co},{t(r['status']) or 'absent'},{t(r['confidence'])},{t(r['overtime_hrs'])}")
    csv = '\n'.join(lines)
    return Response(csv, mimetype='text/csv',
                    headers={'Content-Disposition': f'attachment; filename=attendance-{att_date}.csv'})
